recipes-extended/xen/files/libxl-fixup-cmdline-ops.patch

```diff
@@ -0,0 +1,384 @@
+--- xen-4.6.1.orig/tools/libxl/libxl.h
++++ xen-4.6.1/tools/libxl/libxl.h
+@@ -1171,6 +1171,8 @@ int libxl_domain_remus_start(libxl_ctx *
+ 
+ int libxl_domain_shutdown(libxl_ctx *ctx, uint32_t domid);
+ int libxl_domain_reboot(libxl_ctx *ctx, uint32_t domid);
++int libxl_domain_sleep(libxl_ctx *ctx, uint32_t domid);
++int libxl_domain_hibernate(libxl_ctx *ctx, uint32_t domid);
+ int libxl_domain_destroy(libxl_ctx *ctx, uint32_t domid,
+                          const libxl_asyncop_how *ao_how)
+                          LIBXL_EXTERNAL_CALLERS_ONLY;
+Index: xen-4.6.1/tools/libxl/libxl_types.idl
+===================================================================
+--- xen-4.6.1.orig/tools/libxl/libxl_types.idl
++++ xen-4.6.1/tools/libxl/libxl_types.idl
+@@ -154,6 +154,7 @@ libxl_trigger = Enumeration("trigger", [
+     (4, "INIT"),
+     (5, "RESET"),
+     (6, "S3RESUME"),
++    (7, "HIBERNATE"),
+     ])
+ 
+ libxl_tsc_mode = Enumeration("tsc_mode", [
+@@ -296,6 +297,7 @@ libxl_dominfo = Struct("dominfo",[
+     #
+     # Otherwise set to a value guaranteed not to clash with any valid
+     # LIBXL_SHUTDOWN_REASON_* constant.
++    ("power_command", uint32),
+     ("shutdown_reason", libxl_shutdown_reason),
+     ("outstanding_memkb",  MemKB),
+     ("current_memkb",   MemKB),
+Index: xen-4.6.1/tools/libxl/libxl_utils.c
+===================================================================
+--- xen-4.6.1.orig/tools/libxl/libxl_utils.c
++++ xen-4.6.1/tools/libxl/libxl_utils.c
+@@ -93,6 +93,64 @@ int libxl_name_to_domid(libxl_ctx *ctx,
+     return ret;
+ }
+ 
++int libxl_get_acpi_state(libxl_ctx *ctx, int32_t domid, uint32_t *acpi_state)
++{
++    unsigned long hvm_s_state = 0;
++    int ret;
++    ret = xc_get_hvm_param(ctx->xch, domid, HVM_PARAM_ACPI_S_STATE, &hvm_s_state);
++    if (ret < 0) { 
++        hvm_s_state = 9; //since unsigned, use invalid acpi state
++    }
++    *acpi_state = hvm_s_state;
++    return 0;
++}
++
++int libxl_uuid_to_domid(libxl_ctx *ctx, const char *uuid_in, int32_t *domid)
++{
++    
++    int i, nb_domains;
++    libxl_dominfo *dominfo;
++    libxl_uuid uuid, uuid2;
++    int ret = 0;
++    *domid = -1;
++    libxl_uuid_from_string(&uuid2, uuid_in);
++    dominfo = libxl_list_domain(ctx, &nb_domains);
++    if (!dominfo)
++        return ERROR_NOMEM;
++    
++    for (i = 0; i < nb_domains; i++) {
++        uuid = dominfo[i].uuid;
++        if (libxl_uuid_compare(&uuid, &uuid2) == 0) {
++            *domid = dominfo[i].domid;
++            ret = 0;
++            break;
++        }
++    }
++    free(dominfo);
++    return ret;
++}
++
++int libxl_domid_to_uuid(libxl_ctx *ctx, libxl_uuid **uuid, uint32_t domid_in)
++{
++    int nb_domains, i;
++    uint32_t domid;
++    libxl_dominfo *dominfo;
++
++    dominfo = libxl_list_domain(ctx, &nb_domains);
++
++    for(i = 0; i < nb_domains; i++)
++    {
++        domid = dominfo[i].domid;
++        if (domid == domid_in)
++        {
++            *uuid = &dominfo[i].uuid;
```

Comment:
No.  libxl_list_domain makes a hypercall _xc_domain_getinfolist_, takes that list and calls _xcinfo2xlinfo_, which does a deep-copy/conversion to a libxl_dominfo struct.  So as along as the libxl_dominfo is around, the uuid will be a valid pointer.

That being said, we might have a memory leak here since we don't free dominfo before returning.  We should probably have the caller allocate space for the libxl_uuid, do a deep copy, and free dominfo before returning.